Fix Ul'xzomit roaming and spawn behaviors

DiscipleOfEris opened the pull request:
**_I affirm:_**
- [x] I understand that if I do not agree to the following points by completing the checkboxes my PR will be ignored.
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/LandSandBoat/server/blob/base/CODE_OF_CONDUCT.md).
- [x] I have _**tested my code and the things my code has changed**_ since the last commit in the PR and will test after any later commits.

## What does this pull request do?

Fixes #2668 

Make baby Ul'xzomits spawn normally.

Ul'xzomit babies will now roam just like mother Ul'xzomits when the mother is dead. They will re-attach when the mother respawns, and will re-attach if the mother is alive when they spawn.

Reduce Ul'xzomit respawn time to be in-line with retail.

Fix a bug where Ul'xzomit babies could be immune to physical damage (seemed to be a spawn race condition I couldn't quite figure out).

Changed some defaults of `xi.follow.follow()` params, and added option to disable warping.

## Steps to test these changes

1. Go to Al'taieu.
2. Kill a baby Ul'xzomit and respawn it. Observe it automatically re-attaches.
3. Kill a mother Ul'xzomit. Observe the baby Ul'xzomits start roaming by themselves.
4. Kill a baby Ul'xzomit and respawn it. Observe it starts roaming by itself.
5. Respawn the mother Ul'xzomit. Observe the babies start following it again.
6. Fight a mother Ul'xzomit. Kill a baby Ul'xzomit and respawn it. Observe the baby only links when it gets close.
7. Kill an Ul'xzomit and let it respawn normally. Observe its respawn time is now 5.5 minutes.

## Other thoughts

~~I believe baby Ul'xzomits when following the mother have the `xi.pathFlag.RUN` flag set, despite my not setting it. My observations suggest their follow speed should be the normal roam speed. On retail, if they fall behind the mother they don't catch up unless the mother changes directions.~~

~~Baby Ul'xzomits might be linking *slightly* farther than they do in retail. It's harder to test since the babies move a little too fast.~~

~~The following should probably place a limit on its max pathfinding distance it calculates at one time. If the baby Ul'xzomits are far from the mother, the mother could have moved quite far by the time the baby gets there.~~

Ul'xzomits when roaming seem to occasionally despawn on their own (they do not on retail), perhaps by getting too far from their spawn. This is a general problem with LSB's roaming behavior, and not specific to the Ul'xzomit behavior implemented in this PR.

When a baby Ul'xzomit delatches, it sometimes moves in a straight line quite some distance to return to its "area" which isn't necessarily the same as the mother. This is a general problem in LSB's roaming behavior, and not specific to the Ul'xzomit behavior implemented in this PR.

Ul'aerns, Ul'hpemdes, and Ul'phuabos also seem to have 5.5 minute respawns on retail. ~~I haven't checked Om' or other mobs in Al'taieu, but~~ I suspect most of them have 5.5 minute respawns since Al'taieu is an outdoor zone (though you cannot mount). (See #3630)

DiscipleOfEris committed:
Fix Ul'xzomit roaming and spawn behaviors

Make baby Ul'xzomits spawn normally.

Reduce Ul'xzomit respawn time to be in-line with retail.

Fix a bug where Ul'xzomit babies could be immune to physical damage (seemed to be a spawn race condition I couldn't quite figure out).

Ul'xzomit babies will now roam just like mother Ul'xzomits when the mother is dead. They will re-attach when the mother respawns, and will re-attach if the mother is alive when they spawn.

Changed some defaults of `xi.follow.follow()` params, and added option to disable warping.

DiscipleOfEris committed:
Sanity check compliance

Xaver-DaRed commented on scripts/globals/follow.lua:
```diff
@@ -56,30 +57,59 @@ function addFollower(leader, follower, options)
 
-function removeFollower(leader, follower)
-    local leaderId = leader:getID()
+function cleanupFollower(follower)
```
Since this is a global function, can it be changed to `xi.follow.cleanupFollower`?

DiscipleOfEris replied:
`cleanupFollower()` is a local function. I declared it local at the beginning of the file. That said, it's only called in one place (as a result of refactoring from when I added it) so I'll just roll it into `xi.follow.stopFollowing()`

TeoTwawki replied:
Didn't catch it previously (way back in [august last year](https://github.com/LandSandBoat/server/commit/cb3458a2553f54546b3d2f52e8f53e2fbb406722#diff-840a590675fe79f6b29c82d4bb7eed93d24c0c273a5ea5a5d397aed8690670c9R18)) but I dislike the pattern of functions beginning their lives as nil variables. Which is why nobody noticed this was local and not global.

DiscipleOfEris replied:
For clarity, it's a way to keep functions in local scope but have the function definition after the function definition of a function that will call the local one. Since Lua is a single-pass compiler. I'm still using it for the `onMobDespawn` and `onMobRoamAction` callbacks.

If it's an approach that LSB would prefer to avoid, let me know.

TeoTwawki replied:
Definitely want to avoid.

DiscipleOfEris replied:
Okay, I've re-ordered the functions, as requested.

Xaver-DaRed commented on scripts/globals/follow.lua:
```diff
@@ -56,30 +57,59 @@ function addFollower(leader, follower, options)
+    follower:setMobMod(xi.mobMod.ROAM_COOL, follower:getLocalVar("[follow]ROAM_COOL"))
+end
+
+function xi.follow.clearFollowers(leader)
```
Related to the comment above, can global functions that you edit follow this structure?

xi.follow.functionName = function(params)

Xaver-DaRed commented on scripts/globals/follow.lua:
```diff
@@ -56,30 +57,59 @@ function addFollower(leader, follower, options)
+    end
+end
+
+function xi.follow.stopFollowing(follower)
```
Same here

Xaver-DaRed commented on scripts/zones/AlTaieu/mobs/Ulxzomit.lua:
```diff
@@ -3,75 +3,83 @@
 local entity = {}
 
-entity.onMobSpawn = function(mob)
+function entity.onMobSpawn(mob)
```
We follow the deleted method to name functions. For consistency across  the project. Sorry ><

DiscipleOfEris replied:
No prob. It was a change I overlooked when porting my fix from my own server.

TeoTwawki replied:
page load move where I was typing.

Xaver-DaRed commented on scripts/zones/AlTaieu/mobs/Ulxzomit.lua:
```diff
@@ -3,75 +3,83 @@
-        return
-    end
+function entity.onMobEngaged(mob, target)
+end
```
This function is empty, so it isn't really needed. Pls remove.

Same with any other empty functions, except the main 4 basic ones (spawn, death, despawn, etc)

DiscipleOfEris committed:
Style compliance

Requested changes

DiscipleOfEris committed:
Fix follow `pathFlags` being ignored

Also implement requested change on removing forward-declaration of local functions.

DiscipleOfEris commented:
I figured out why my lack of `pathFlag.RUN` was being ignored. I never actually told the call to `pathTo()` to use the configured option...

DiscipleOfEris commented:
While the `pathFlag` is taken care, it doesn't seem to be possible to limit the length of the `pathTo`. I could see adding a new function to the pathing AI and a new Lua binding, that prunes points beyond a max range (as opposed to `pathInRange` which sets a min range).

The other option would be to add a check to `onMobRoamAction` and just clear the current path when the follower has moved more than the max distance, so it re-paths.

The final option would be to just leave it as-is.

DiscipleOfEris commented:
I think I'll go with a `maxSecondsBetweenPathing` option for `follow()`, that if set will record the time of each `pathTo()` and then clear the previous `pathTo()` after that many seconds, to repath.

DiscipleOfEris committed:
Merge remote-tracking branch 'upstream/base' into fix-ulxzomit-behaviors

DiscipleOfEris committed:
Add `forceRepathInterval` option to `xi.follow`

This option will cause the follower to cancel its previous path early, to compensate for a faraway leader that might have moved long before the follower arrives.

Also, change default value of the `warpDistance` option to be disabled by default. And some minor refactoring for style and to reduce repetition.

DiscipleOfEris commented:
@zach2good @claywar @Xaver-DaRed Ready for review.

Xaver-DaRed reviewed (approved):
Looks sane enough to me.

Kind of weird that ONLY Ul'xzomit use the follow logic currently.
For instance, there is this quadav NM in Pashow Marshlands that has 2 scorts.

Out of scope of this PR, of course.

DiscipleOfEris commented:
Yes, I coded this with the intent that the follow logic could be used by other mobs.

DiscipleOfEris commented:
@zach2good @TeoTwawki @Xaver-DaRed Is there anything blocking this being merged?
